### ISSUE TITLE
loss value from slim.learning.train

### ISSUE THREAD

**Issue by zhang-jian** [tensorflow/magenta#607]
Hi,

I am getting outputs:
INFO:tensorflow:global step 2511: loss = **2510.0000** (1.71 sec/step)
INFO:tensorflow:global step 2511: loss = **2511.0000** (0.93 sec/step)
INFO:tensorflow:global step 2512: loss = **2512.0000** (0.91 sec/step)
INFO:tensorflow:global step 2513: loss = **2513.0000** (0.94 sec/step)
INFO:tensorflow:global step 2514: loss = **2514.0000** (0.94 sec/step)
INFO:tensorflow:global step 2515: loss = **2515.0000** (0.94 sec/step)

It seems that current train_op doesn't return total loss as it is required by slim.learning.train, or I am wrong?

Thanks

**Comment by cghawthorne** [comment]
I think this is a question about the nsynth code. @adarob or @jesseengel could you take a look?

**Comment by jesseengel** [comment]
It looks like the global step is being reported as the loss. Can you give the command line call that you used to produce this output and magenta commit number?

**Comment by Invisibility** [comment]
I encountered similar issue. The command line I used was:

> bazel run //magenta/models/nsynth/wavenet:train -- --train_path=/ml/data/nsynth/tfrecord/nsynth-train.tfrecord --logdir=/ml/tensorflow/magenta/log

The commit number is https://github.com/tensorflow/magenta/commit/935438f18f020b31e6438889dbcfbd922595e9f7.

The train_path file was downloaded from http://download.magenta.tensorflow.org/datasets/nsynth/nsynth-train.tfrecord.

I am using ubuntu 14.04, with a python 2.7.6 virtual environment and the following packages:

    audioread==2.1.5
    backports.weakref==1.0rc1
    bleach==1.5.0
    Cython==0.26
    decorator==4.1.2
    funcsigs==1.0.2
    html5lib==0.9999999
    joblib==0.11
    librosa==0.5.1
    Markdown==2.2.0
    mock==2.0.0
    numpy==1.13.1
    pbr==3.1.1
    protobuf==3.3.0
    resampy==0.1.5
    scikit-learn==0.18.2
    scipy==0.19.1
    six==1.10.0
    tensorflow==1.2.1
    tensorflow-tensorboard==0.1.2
    Werkzeug==0.12.2

I compiled my tensorflow package with CUDA 8.0 and CUDNN 6.0. My GPU is TITAN X (Pascal).

**Comment by auzxb** [comment]
I meet the same problem. How to solve it?

**Comment by 493238731** [comment]
I meet the same problem.And the loss always is NaN in the tensorboard, after 50k step using default params.
Maybe it uses batch_size = 1 in default params to train unstable?

**Comment by Invisibility** [comment]
Based on @zhang-jian observation that train_op isn't the right operation, and the code in baseline model (baseline/train.py and baseline/models/ae.py), I made the following change in wavenet/train.py:

Instead of:

      train_op = opt.minimize(
          loss,
          global_step=global_step,
          name="train",
          colocate_gradients_with_ops=True)

I change it to:

      train_op = slim.learning.create_train_op(
          loss,
          opt,
          global_step=global_step,
          colocate_gradients_with_ops=True)

This appears to make loss output not the same as global step. 

However, my training was still not successful (forgot whether it has NaN or the loss not converging). It may be because I didn't have enough GPU memory to run with 32 worker replicas and 32 total_batch_size as mentioned in the parameter descriptions.

**Comment by 493238731** [comment]
You can see [#853](https://github.com/tensorflow/magenta/issues/853)

**Comment by Invisibility** [comment]
Thank you @493238731. I am rerunning the training, and it seems to run further than before. 

Based on the suggestion in #853, I updated the function in utils.py as follows:

  def mu_law(x, mu=255, int8=False):
    """A TF implementation of Mu-Law encoding.

    Args:
      x: The audio samples to encode.
      mu: The Mu to use in our Mu-Law.
      int8: Use int8 encoding.

    Returns:
      out: The Mu-Law encoded int8 data.
    """
    out = tf.clip_by_value(x, -1, 0.999999999)
    out = tf.sign(out) * tf.log(1 + mu * tf.abs(out)) / np.log(1 + mu)
    out = tf.floor(out * 128)
    if int8:
      out = tf.cast(out, tf.int8)
    return out

**Comment by 493238731** [comment]
@Invisibility Your code maybe wrong? You should run 'out = tf.clip_by_value(x, -1, 0.9999999)' after between `out = tf.sign(out) * tf.log(1 + mu * tf.abs(out)) / np.log(1 + mu)` and `out = tf.floor(out * 128)`
.Because the x is in [0, 2^16 -1] not in [-1, 1].
In addition, the`0.999999999` is too small to treated as 1. in tf.float32/np.float32 mode, you can change it bigger e.g. 0.99, 0.999

**Comment by Invisibility** [comment]
@493238731 Thanks for the notes. I agree that 0.99 or 0.999 value and moving the statement to the new location as you suggested are better due to the float32 issue. My last run failed probably because of the float32 issue. I am rerunning with what you suggested.

I do think that the input x to the mu_law function is in [-1, 1], because the [µ-law formula](https://en.wikipedia.org/wiki/%CE%9C-law_algorithm) requires that.
